File: DONGUK/프로그래머스_WEEK 1/오픈채팅방.py
# 시작: 22:48
# 종료: 23:00
def solution(record):
    answer = []
    
    user_dict = {} # uid: name
    for data in record:
        data = data.split(' ')
        
        status, uid= data[0], data[1]
        
        if status == 'Enter' or status == 'Change':
            name = data[2]
            
            if status == 'Enter':
                user_dict[uid] = name
            
            elif status == 'Change':
                user_dict[uid] = name
    
    
    
    for data in record:
        data = data.split(' ')
        status, uid= data[0], data[1]
        
        if status == 'Enter':
            ans = "{}님이 들어왔습니다.".format(user_dict[uid])
            answer.append(ans)
            
        elif status == 'Leave':
            ans = "{}님이 나갔습니다.".format(user_dict[uid])
            answer.append(ans)
            
    return answer

File: JUNHEE/프로그래머스_옛날문제/오픈채팅방.cpp
#include <string>
#include <vector>
#include <sstream>
#include <map>

using namespace std;

struct rcd{
    string status;
    string id;
};

vector<string> solution(vector<string> record) {
    vector<string> answer;
    map<string, string> ids;
    vector<rcd> records;
    stringstream ss;
    for(int i = 0; i < record.size();i++){
        ss.str(record[i]);
        string t_status, t_id, t_name;
        ss >> t_status >> t_id >> t_name;
        if(t_status == "Change"){
            ids[t_id] = t_name;
        }
        else if(t_status == "Enter") {
            struct rcd t_rcd = {t_status, t_id};
            ids[t_id] = t_name;
            records.push_back(t_rcd);
        }
        else {
            struct rcd t_rcd = {t_status, t_id};
            records.push_back(t_rcd);
        }
        ss.clear();
    }
    for(auto k : records){
        if(k.status == "Enter"){
            string t_answer = ids[k.id] + "님이 들어왔습니다.";
            answer.push_back(t_answer);
        }
        else{
            string t_answer = ids[k.id] + "님이 나갔습니다.";
            answer.push_back(t_answer);
        }
    }
    return answer;
}
